lib/firmware/fsm_msg_nano.h

```diff
@@ -0,0 +1,309 @@
+        representative_pk);
+    memcpy(balance_be, msg->balance.bytes, sizeof(balance_be));
+    bn_from_bytes(balance_be, sizeof(balance_be), &balance);
+
+    nano_hash_block_data(account_pk, parent_hash, link,
+                         representative_pk, balance_be,
+                         block_hash);
+
+    if (invalid) {
+        fsm_sendFailure(FailureType_Failure_Other, _("Block data invalid"));
+        layoutHome();
+        return;
+    }
+
+    // Some additional sanity checks now that balance values are known
+    char recipient_address[MAX_NANO_ADDR_SIZE];
+    memset(recipient_address, 0, sizeof(recipient_address));
+
+    char representative_address[MAX_NANO_ADDR_SIZE];
+    strlcpy(representative_address, msg->representative, sizeof(representative_address));
+    if (msg->has_parent_block) {
+        if (!strncmp(representative_address,
+                     msg->parent_block.representative,
+                     sizeof(representative_address))) {
+            // Representative hasn't changed, zero it out
+            memset(representative_address, 0, sizeof(representative_address));
+        }
+    }
+    if (strlen(representative_address) > 0) {
+        nano_truncate_address(coin, representative_address);
+    }
+
+    bool is_send = false;
+    if (bn_is_less(&balance, &parent_balance)) {
+        is_send = true;
+        bn_subtract(&parent_balance, &balance, &balance_delta);
+    } else {
+        bn_subtract(&balance, &parent_balance, &balance_delta);
+    }
+
+    if (bn_is_zero(&balance_delta)) {
+        // Balance can only remain unchanged when the account exists already
+        invalid |= !msg->has_parent_block;
+    } else if (is_send) {
+        // For sends make fill out the link_recipient (or generate it from link bytes)
+        if (msg->has_link_recipient) {
+            strlcpy(recipient_address, msg->link_recipient, sizeof(recipient_address));
+        } else {
+            nano_get_address(link,
+                coin->nanoaddr_prefix, strlen(coin->nanoaddr_prefix),
+                recipient_address, sizeof(recipient_address));
+        }
+        nano_truncate_address(coin, recipient_address);
+    } else {
+        // For receives make sure that the link_hash was specified and that it's
+        invalid |= !msg->has_link_hash;
+        uint8_t link_empty[sizeof(link)];
+        memset(link_empty, 0, sizeof(link_empty));
+        invalid |= !memcmp(link_empty, link, sizeof(link));
+    }
+
+    if (invalid) {
+        fsm_sendFailure(FailureType_Failure_Other, _("Block data invalid"));
+        layoutHome();
+        return;
+    }
+
+    // Determine what type of prompt to display
+    bool needs_confirm = true;
+    bool is_transfer = false;
+    if (strlen(representative_address) > 0) {
+        // always confirm representative change
+    } else if (!is_send) {
+        // don't bother confirming pure receives
+        needs_confirm = false;
+    } else if (msg->has_address_type &&
```

Comment:
I think this whole `else if` for transfers ought to just be `} else if (msg->link_recipient_n_count) {`, since the path gets checked in the *_to_string helpers.

Reply:
I always felt like the way that's handled for UTXO coins was a bit odd, so no need to replicate that strange behavior here.

Reply:
For exchange, if/when we add that, it'd be a call to `run_policy_compile_output` here, and set `needs_confirm = false;` if it succeeds.

Reply:
Yeah, was trying to mimic the existing implementations, totally agree that address_type is kind of redundant here. This will also means I can get rid of the NanoSignTx.address_type which I didn't really like in the protocol def.

Reply:
Changed with 2bcfe03